### PR TITLE
Hw10 ser des

### PR DESCRIPTION
**Pull request by GevorgeX**

### CODE REVIEW

File: Utils/SerDes/SerDes.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <cstring>
+
+template<class T>
+void serialization(const std::string& file, const T* obj ) {
+    int fd = open(file.c_str(), O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR);
+
+    auto bytes = reinterpret_cast<const char*>(obj);
+    write(fd, bytes, sizeof(T));
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: Use open/truncate/mmap/munmap.

File: Utils/SerDes/SerDes.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include <fcntl.h>
+#include <string>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <cstring>
+
+template<class T>
+void serialization(const std::string& file, const T* obj ) {
+    int fd = open(file.c_str(), O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR);
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: Check the return values of system calls.

File: Utils/SerDes/SerDes.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    int fd = open(file.c_str(), O_RDONLY);
+    struct stat st{};
+    fstat(fd, &st);
+
+    auto t = mmap(0, st.st_size, PROT_READ, MAP_SHARED,
+        fd, 0);
+
+    if (t == MAP_FAILED) {
+        perror("mmap");
+        return;
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: Imagine the user calls your function:
```c++
SomeClass instance{};
deserialization("obj.dat", &instance);
```

How would the user be sure that the function succeeded and `instance` is a valid, deserialized object ?

File: Utils/SerDes/SerDes.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    std::memcpy(obj, t, st.st_size);
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: Why are you sure that the `t` is a mmap of a file with valid object serialization of type T ?

File: Utils/SerDes/SerDes.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    int fd = open(file.c_str(), O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR);
+
+    auto bytes = reinterpret_cast<const char*>(obj);
+    write(fd, bytes, sizeof(T));
+
+    close(fd);
+}
+
+template<class T>
+void deserialization(const std::string& file, T* obj) {
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: The use of a pointer forces you to check it against `nullptr` which you didn't do. Make it `T&`.

File: Utils/SerDes/SerDes.h (hunk 4 above)
Comment (MikayelSmbatian) [reply Utils/SerDes/SerDes.h]: Also, it is more safe that you copy exactly `sizeof(T)` instead of the whole file. There is no guarantee that the file size has the exact `sizeof(T)` value. It is better have a corrupted object instead of corrupted memory.

File: Demo/SerDes/SerDesDemo.cpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+struct  ST {
+    int a;
+    float b;
+    char c;
+
+    ST(){}
+    ST(int a, float b, char c): a(a), b(b), c(c) {}
+};
+
+
```
Comment (MikayelSmbatian) [inline Demo/SerDes/SerDesDemo.cpp:16]: Remove extra newline.

File: Demo/SerDes/SerDesDemo.cpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+int main() {
+    ST obj(19,-145.3, 'g');
+    serialization("temp.txt",&obj);
+
+    ST another_obj;
+    deserialization("temp.txt",&another_obj);
+
+    std::cout << another_obj.a << std::endl;
```
Comment (MikayelSmbatian) [inline Demo/SerDes/SerDesDemo.cpp]: instead of printing check those two objects via `std::memcmp` or compare them componentwise and print the result.

Comment (MikayelSmbatian) [reply Demo/SerDes/SerDesDemo.cpp]: Add more unit tests with edge cases.

**Review by MikayelSmbatian — CHANGES_REQUESTED** [review]

File: Demo/SerDes/SerDesDemo.cpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        serialization("temp3.txt",obj);
+
+        ST another_obj;
+        deserialization("temp3.txt",another_obj);
+
+        std::cout<<"First field: " << obj.a <<" == "<< another_obj.a << std::endl;
+        std::cout<<"Second field: "<< obj.b <<" == "<< another_obj.b << std::endl;
+        std::cout<<"Third field: " << obj.c <<" == "<< another_obj.c << std::endl;
+    }
+}
```
Comment (MikayelSmbatian) [inline Demo/SerDes/SerDesDemo.cpp:69]: Add a test case for huge data serialize/deserialize.

Comment (MikayelSmbatian) [reply Demo/SerDes/SerDesDemo.cpp:69]: Take a struct with `char m_data[1024*1024*1024]` and fill it with random data.

File: Utils/SerDes/SerDes.h [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+#include <cstring>
+
+template<class T>
+void serialization(const std::string& file, const T& obj ) {
+    int fd = open(file.c_str(), O_RDWR | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);
+    if (fd == -1) {
+        throw std::runtime_error("Failed to open file");
+    }
+
+    size_t length = sizeof(T);
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: use `constexpr`

File: Utils/SerDes/SerDes.h [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+#include <fcntl.h>
+#include <string>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <cstring>
+
+template<class T>
+void serialization(const std::string& file, const T& obj ) {
+    int fd = open(file.c_str(), O_RDWR | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: Remind me to explain RAII.

**Review by MikayelSmbatian — CHANGES_REQUESTED** [review]

File: Utils/SerDes/SerDes.h (hunk 10 above)
Comment (GevorgeX) [reply Utils/SerDes/SerDes.h]: OK

File: Utils/SerDes/SerDes.h [hunk 11]
```diff
@@ -0,0 +1,55 @@
+#pragma once
+#include <fcntl.h>
+#include <string>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <cstring>
+
+template<class T>
+void serialization(const std::string& file, const T& obj ) {
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: This function should not be compiled for non-trivially copyable types.

File: Utils/SerDes/SerDes.h [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    std::memcpy(t, &obj, length);
+
+    munmap(t, length);
+    close(fd);
+}
+
+template<class T>
+void deserialization(const std::string& file, T& obj) {
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: This function should not be compiled for non-trivially copyable types.

File: Demo/SerDes/SerDesDemo.cpp [hunk 13]
```diff
@@ -0,0 +1,52 @@
+#include <iostream>
+#include <SerDes/SerDes.h>
+
+struct ST {
+    unsigned int a;
+    long long b;
+    char c;
+
+    ST(){}
+    ST(unsigned int a, long long b, char c): a(a), b(b), c(c) {}
```
Comment (MikayelSmbatian) [inline Demo/SerDes/SerDesDemo.cpp]: This violates the CPP Coding Style "_Class and Struct Definitions: 2. Constructor Initialization Lists_".

**Review by MikayelSmbatian — CHANGES_REQUESTED** [review]

File: Utils/SerDes/SerDes.h (hunk 11 above)
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: Use dedicated `std::filesystem::path` instead of `std::string`.

File: Utils/SerDes/SerDes.h [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+    auto t = mmap(nullptr,length, PROT_READ, MAP_PRIVATE,
+                  fd, 0);
+
+    if (t == MAP_FAILED) {
+        close(fd);
+        throw std::runtime_error("Failed to mmap file");
+    }
+
+    std::memcpy(&obj, t, length);
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: What if the provided path doesn't contain an object previously serialized by `serialize` function ?

**Review by MikayelSmbatian — CHANGES_REQUESTED** [review]

File: Demo/SerDes/SerDesDemo.cpp [hunk 15]
```diff
@@ -0,0 +1,69 @@
+#include <iostream>
+#include <SerDes/SerDes.h>
+
+struct ST {
+    unsigned int a;
+    long long b;
+    char c;
+
+    ST() : a(0), b(0), c('\0') {}
```
Comment (MikayelSmbatian) [inline Demo/SerDes/SerDesDemo.cpp]: use `ST() = default;` instead.

File: Demo/SerDes/SerDesDemo.cpp [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+        Mec obj{};
+        for (size_t i = 0;i<1024*1024;i++) {
+            obj.m_data[i] = 'G';
+        }
+        serialization("temp4.txt",obj);
+
+        Mec another_obj{};
+        deserialization("temp4.txt",another_obj);
+
+        std::cout<<"First field: " << obj.m_data[1025] <<" == "<< another_obj.m_data[1025] << std::endl;
```
Comment (MikayelSmbatian) [inline Demo/SerDes/SerDesDemo.cpp]: Use `std::memcmp`.

File: Utils/SerDes/SerDes.h [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <cstring>
+#include <filesystem>
+
+#define MAGIC_NUMBER ('R' + 'O' + 'B')
+class SmartMap {
+    unsigned char* data;
+    size_t length;
+    public:
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: Align access specifiers with the `class` keyword
```c++
class <name> {

public:

}; 
```

File: Utils/SerDes/SerDes.h [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+#include <string>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <cstring>
+#include <filesystem>
+
+#define MAGIC_NUMBER ('R' + 'O' + 'B')
+class SmartMap {
+    unsigned char* data;
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: use `std::byte*` instead.

File: Utils/SerDes/SerDes.h [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+    unsigned char* get() const {
+        return data;
+    }
+};
+
+template<class T>
+void serialization(const std::filesystem::path& file, const T& obj ) {
+    static_assert(std::is_trivially_copyable_v<T>);
+
+    int fd = open(file.c_str(), O_RDWR | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: You should have implemented FileDescriptor RAII alongside with the SmartMap.

File: Utils/SerDes/SerDes.h [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+    int fd = open(file.c_str(), O_RDONLY);
+    if (fd == -1) {
+        throw std::runtime_error("Failed to open file");
+    }
+    constexpr size_t length = sizeof(T);
+
+    SmartMap t(nullptr,length+1, PROT_READ, MAP_PRIVATE,fd, 0);
+
+    if(*t.get() != MAGIC_NUMBER) {
+        throw std::runtime_error("Failed to deserialize file");
```
Comment (MikayelSmbatian) [inline Utils/SerDes/SerDes.h]: More descriptive log would be better.

**Review by MikayelSmbatian — APPROVED** [review]
